### PR TITLE
ADR #6: Completed Action Logger and Monetary Resource Usage

### PR DESCRIPTION
**Pull request by travistakai**
Creating an ADR to introduce the new Completed Action Logger service for streaming action results to an external service. This will allow for analyzing action results in real-time as well as providing a means for persisting this data. Also introduce a new form of metadata named MonetaryResourceUsage which will be embedded in the auxiliary_metadata field.

### CODE REVIEW

File: 0006-operation-logging-and-monetary-resource-usage.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+The protocol will be used by bb_worker to stream details of completed actions, known as CompletedActions, to an external logging service:
+
+```protobuf
+service CompletedActionLogger {
+  // Send a CompletedAction to another service as soon as a build action has
+  // completed. Receiving a message from the return stream indicates that the
+  // client successfully received the CompletedAction.
+  rpc LogCompletedActions(stream CompletedAction)
+      returns (stream google.protobuf.Empty);
```
Comment (EdSchouten) [inline 0006-operation-logging-and-monetary-resource-usage.md:31]: Could you add a sentence here explaining what's the role of the return stream? For example:

> For every CompletedAction that the service processes successfully, an empty message is streamed back to the client.

**Review by EdSchouten — APPROVED** [review]
Looks great! :rocket:

File: 0006-operation-logging-and-monetary-resource-usage.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+# Completed Action Logger Service
+
+The protocol will be used by bb_worker to stream details of completed actions, known as CompletedActions, to an external logging service:
+
+```protobuf
+service CompletedActionLogger {
+  // Send a CompletedAction to another service as soon as a build action has
+  // completed. Receiving a message from the return stream indicates that the
+  // client successfully received the CompletedAction.
```
Comment (EdSchouten) [inline 0006-operation-logging-and-monetary-resource-usage.md:29]: s/client/service/

**Comment by Qinusty** [comment]
Looking forward to seeing this combined with `target_id` from [this PR](https://github.com/bazelbuild/remote-apis/pull/186/files#diff-c43847ba40caac956777616525d7261d09782e0f5f07c5a53afbbcfcb4fd338eR1700)!

Are there any thoughts towards how workers will be configured to estimate cost based on resource consumption? Is this something the individual workers should be calculating or would simply reporting their resource consumption be more flexible for cost calculation in single service?

**Comment by EdSchouten** [comment]
> Looking forward to seeing this combined with `target_id` from [this PR](https://github.com/bazelbuild/remote-apis/pull/186/files#diff-c43847ba40caac956777616525d7261d09782e0f5f07c5a53afbbcfcb4fd338eR1700)!

Yeah, pretty awesome! I believe Bazel master is already setting this properly. bb_scheduler and bb_browser also print that info nowadays. \o/

> Are there any thoughts towards how workers will be configured to estimate cost based on resource consumption? Is this something the individual workers should be calculating or would simply reporting their resource consumption be more flexible for cost calculation in single service?

We've already added a basic implementation of this part of the ADR to bb-remote-execution: https://github.com/buildbarn/bb-remote-execution/commit/c826ba0d7234cde2337c79f9719dd0fe21c88c8f

Right now it's basically a setting in the worker configuration, where you specify the cost per second. We can always extend these facilities if more complex ways of computing costs are desired.

**Comment by Qinusty** [comment]
> Yeah, pretty awesome! I believe Bazel master is already setting this properly. bb_scheduler and bb_browser also print that info nowadays. \o/

 Nice, I'll take a look! 

> Right now it's basically a setting in the worker configuration, where you specify the cost per second. We can always extend these facilities if more complex ways of computing costs are desired.

Looks useful! From this I can imagine it would be useful to combine this with metadata for time spent computing/fetching/uploading for breakdowns of cost relating to IO vs compute. Since the runner is blocked from compute during the IO operations this should be straight forward using the existing metadata.

**Comment by ulrfa** [comment]
Interesting with interface for historical data!

What do you think about also streaming:

1. [tool_invocation_id,](https://github.com/bazelbuild/remote-apis/blob/0cd22f7b466ced15d7803e8845d08d3e8d2c51bc/build/bazel/remote/execution/v2/remote_execution.proto#L1760-L1762) in order to group low level actions into high level bazel invocations that users perform.

2. Metadata set by users as headers in gRPC context, e.g. via bazel's [--remote_exec_header](https://docs.bazel.build/versions/master/command-line-reference.html#flag--remote_exec_header)="key=value". In order to allow users to attach (arbitrary?) metadata such USER=Alice, HOST=machine123, TYPE_OF_BUILD=integration.

Related to: buildbarn/bb-remote-execution/pull/72

**Comment by travistakai** [comment]
> Interesting with interface for historical data!
> 
> What do you think about also streaming:
> 
> 1. [tool_invocation_id,](https://github.com/bazelbuild/remote-apis/blob/0cd22f7b466ced15d7803e8845d08d3e8d2c51bc/build/bazel/remote/execution/v2/remote_execution.proto#L1760-L1762) in order to group low level actions into high level bazel invocations that users perform.

The [auxiliary_metadata](https://github.com/bazelbuild/remote-apis/blob/master/build/bazel/remote/execution/v2/remote_execution.proto#L954) field embedded in the historical_execute_response message will already include the tool_invocation_id since the the RequestMetadata info is included in the HistoricalExecuteResponses by default.

> 2. Metadata set by users as headers in gRPC context, e.g. via bazel's [--remote_exec_header](https://docs.bazel.build/versions/master/command-line-reference.html#flag--remote_exec_header)="key=value". In order to allow users to attach (arbitrary?) metadata such USER=Alice, HOST=machine123, TYPE_OF_BUILD=integration.
> 
> Related to: [buildbarn/bb-remote-execution/pull/72](https://github.com/buildbarn/bb-remote-execution/pull/72)

This seems like it makes sense to add the remote_exec_header key/value pairs to the auxiliary_metadata field. Does the remote_exec_header data get attached to the gRPC headers in the same way authentication info would be?

**Comment by EdSchouten** [comment]
> 2. Metadata set by users as headers in gRPC context, e.g. via bazel's [--remote_exec_header](https://docs.bazel.build/versions/master/command-line-reference.html#flag--remote_exec_header)="key=value". In order to allow users to attach (arbitrary?) metadata such USER=Alice, HOST=machine123, TYPE_OF_BUILD=integration.

I’m not convinced that’s good use of this command line flag. Note that this flag sets HTTP headers. You’re basically mixing in higher level metadata with transport-level things like authentication.

A month or so ago there were discussions within the remote execution working group about how to propagate such metadata. Please work with them to get proper facilities added for doing that.

**Pull request merged**